### PR TITLE
Add SCE merging

### PR DESCRIPTION
**Pull request by yangzy39**
This pull request introduces a novel method for large language model (LLM) merging, leveraging the power of [SCE](https://arxiv.org/abs/2408.07990). SCE incorporates adaptive matrix-level merging weights derived from parameter variance analysis.

**Key Features of SCE:**

- **Selective Attention:** SCE **s**elects high-variance elements across all delta parameters within each parameter matrix. This focuses the merging process on the most informative parts of the models.
- **Adaptive Weighting:**  SCE **c**alculates matrix-level merging weights based on the sum of squares of elements in the delta parameters. This ensures that models with more significant parameter variations contribute more to the fused model.
- **Sparsification:** Similar sign election process in TIES, SCE **e**rases minority elements.

**Improved LLM Performance through SCE:**

We have demonstrated the effectiveness of SCE in our [FuseChat (https://arxiv.org/abs/2408.07990)](https://arxiv.org/abs/2408.07990) paper. Recent days, we  further employ SCE to integrate multiple open-source o1-like LLMs into a unified model. [FuseO1](https://huggingface.co/blog/Wanfq/fuseo1-preview) involves two primary merging strategies:

- **Long-Long Reasoning Merging**: This approach involves model fusion across LLMs that utilize long-CoT reasoning, with the goal of enhancing long-CoT reasoning capabilities. The resulted [FuseAI/FuseO1-DeepSeekR1-QwQ-SkyT1-32B-Preview](https://huggingface.co/FuseAI/FuseO1-DeepSeekR1-QwQ-SkyT1-32B-Preview) achieves a Pass@1 accuracy of **74.0 on AIME24**,  demonstrating significant performance improvements compared to the OpenAI o1-preview (44.6) and OpenAI o1-mini (63.4), even approaching OpenAI o1 (79.2).
- **Long-Short Reasoning Merging**: This approach involves model fusion between long-CoT and short-CoT LLMs, aiming to improve reasoning capabilities in both long and short reasoning processes. The resulted [FuseAI/FuseO1-DeepSeekR1-Qwen2.5-Instruct-32B-Preview](https://huggingface.co/FuseAI/FuseO1-DeepSeekR1-Qwen2.5-Instruct-32B-Preview) is capable of utilizing both long and short reasoning processes and demonstrates relatively strong performance in long reasoning tasks.

We have run the pre-commit hook locally to standardize the formatting, please let us know if you require any modifications or clarifications for this pull request.

### CODE REVIEW

**Comment by cg123** [comment]
Looks great! Thanks for the PR.

**Comment by ehartford** [comment]
I don't see it here:

https://github.com/arcee-ai/mergekit/tree/main/examples

**Review by ehartford — COMMENTED** [review]
@yangzy39 

o1 pro mentioned a potential gotcha:

Below are the most important points to notice when comparing this reference code to the description of SCE in the paper. In particular, pay close attention to how the “erase mask” is computed in `get_erase_mask` versus what the paper says about discarding elements with minority/majority sign.

---

## 1. The key “majority sign” logic may not be doing what you expect

In the code:

```python
def get_erase_mask(
    delta: torch.Tensor,
    mask_dtype: Optional[torch.dtype] = None,
):
    # delta has shape (num_models, ...)

    sign = delta.sign().to(mask_dtype)  # shape (num_models, ...)
    sign_weight = delta.sum(dim=0)      # shape (...)
    majority_sign = (sign_weight >= 0).to(mask_dtype) * 2 - 1
    return sign == majority_sign        # shape (num_models, ...)
```

- `sign_weight = delta.sum(dim=0)` sums up the **raw values** of all deltas at each coordinate, *not* the sign of the deltas. 
- Then `majority_sign` is determined by the sign of that sum. 
- If that net sum is ≥ 0, `majority_sign` is +1; otherwise −1.

**What this means in practice:**  

- An element whose sign is very large in magnitude can “flip” the sum to positive or negative—even if most models have the *opposite* sign.  
- So this is effectively discarding elements that oppose the **dominant (summed) direction**, *weighted by magnitude*, rather than discarding elements whose sign is in the **strict minority** of the *count* of models.

Depending on exactly how the paper defines “discarding changes whose sign is not consistent with the majority of the participating models,” this can be a discrepancy. Often, “majority sign” is taken to mean: 

> **Count** how many models have a positive update versus how many have a negative update, and pick whichever sign has strictly more models behind it.

If that was the intended logic, one would do:

```python
sign_weight = (delta.sign()).sum(dim=0)  # sum of ±1 across models
majority_sign = (sign_weight >= 0).to(mask_dtype)*2 - 1
```

(This ensures that each model’s sign contributes equally, independent of magnitude.)

### Does it match the paper?

1. **If** the paper specifies that “minority direction” is determined purely by counting how many models are positive vs. negative for each coordinate, **then this code is slightly off**—it uses the sign of the sum of raw deltas, not the majority of signs.
2. **If** the paper instead implies that the aggregator discards elements that oppose the net effect (i.e., sign of the sum) for each coordinate, then this code matches that interpretation.

Because the docstring says “Erase elements with minority directions,” and because the typical meaning of “minority sign” is about *counting how many models are in one direction vs. the other*, the safer assumption is that the intended operation is counting. In that case, you’d want `sign_weight = sign.sum(dim=0)`.

---

## 2. The top-k masking by variance is fairly straightforward

```python
if self.select_topk < 1:
    mask = get_sce_mask(deltas, self.select_topk, mask_dtype)
    mask_expanded = mask.unsqueeze(0).expand_as(deltas)
    deltas = deltas * mask_expanded
```

```python
def get_sce_mask(
    deltas: torch.Tensor,
    density: float,
    mask_dtype: Optional[torch.dtype] = None,
):
    # variance along the "model" dimension
    variance = torch.var(deltas, dim=0, unbiased=False)

    # number of non-zero positions
    non_zero_positions_count = torch.count_nonzero(variance)

    # the number of top elements to select
    k = int(abs(density) * non_zero_positions_count)

    # short-circuit if k=0
    mask = torch.zeros_like(variance, dtype=mask_dtype)
    if k == 0:
        return mask

    # pick top-k variance coordinates
    topk_indices = torch.topk(variance.view(-1), k=k, largest=True).indices
    mask.view(-1)[topk_indices] = 1
    return mask
```

1. The code computes the variance across the model dimension and picks the top-`k` fraction of coordinates with highest variance. 
2. It zeroes out everything else. 
3. This logic is consistent with “selecting top τ% elements with highest variance.”  
   - Note that `variance` is always nonnegative, so `.abs()` is a no-op; `.abs()` can safely be omitted.  
   - After that, only “high-variance” coordinates remain nonzero.  

This step appears consistent with the “Selective coordinate” or “pruning by variance” approach typically described in SCE-like methods.

---

## 3. Weighting by average squared magnitude

```python
def get_sce_weight(deltas):
    # average squared magnitude for each delta
    weights = [torch.sum(delta**2).item() / delta.numel() for delta in deltas]

    sum_weights = sum(weights)
    if sum_weights == 0:
        # fallback: uniform
        weights = [1.0 / len(weights)] * len(weights)
    else:
        # normalize so that sum of weights == 1
        weights = [w / sum_weights for w in weights]

    return weights
```

- Each model’s delta is weighted in proportion to its mean squared magnitude.  
- This is a straightforward approach to give models with “larger changes” more or less relative weight.  
- After computing this, the code merges by  
  \[
    \text{mixed\_delta}[j] \;=\; \frac{\sum_{m=1}^M \bigl(\text{weights}_m \cdot \delta_{m,j} \cdot \mathbf{1}(\text{sign}(\delta_{m,j}) = \text{majority\_sign}_j)\bigr)}
                                  {\sum_{m=1}^M \bigl(\text{weights}_m \cdot \mathbf{1}(\text{sign}(\delta_{m,j}) = \text{majority\_sign}_j)\bigr)} 
  \]
  (plus the top-k-variance masking, if applicable).

This part is typically how “SCE weighting” is combined with sign consistency in many coordinate-ensemble methods. It looks correct *provided you genuinely want the sum-of-deltas sign approach described above.*

---

## 4. Int8 masking

```python
mask_dtype = torch.int8 if self.int8_mask else base.dtype
```

This is just an optimization detail (storing the mask in `int8` vs. the same dtype as `base`). It should not affect correctness unless the code later tries to do floating-point operations on the mask. The code simply multiplies the deltas by the mask, so it should be safe.

---

## Summary of the main caution

**Everything in the code is consistent** with:
1. Top-`k%` variance-based selection (if `select_topk < 1.0`).
2. Weighting each delta by its average squared magnitude.
3. Keeping only those model deltas whose sign matches the sign of the *sum of deltas* at that coordinate (rather than strictly the “majority of sign counts”).

If the paper’s SCE description truly says “discard the minority sign in terms of *counting how many are positive vs negative*,” then **you should replace**:

```python
sign_weight = delta.sum(dim=0)
```
with
```python
sign_weight = delta.sign().sum(dim=0)
```
in `get_erase_mask()`. That is the biggest logical difference one commonly sees in “majority sign” vs “dominant direction” code.  

Otherwise, if the paper’s text clarifies that SCE discards the sign that “goes against the net direction of the sum,” the reference code already matches that.  

Given the docstring “Erases elements with minority directions,” most readers interpret “minority direction” as “the fewer models are negative/positive.” However, the current code effectively picks the sign of the *largest net sum*, which can differ if one model’s updates have large magnitude compared to others.  

So double-check the exact definition in your paper or in your method. If you intended “majority sign by count of negative vs. positive deltas,” then that is the one fix you should make. Otherwise, the rest of the code is generally correct for a “dominant direction aggregator.”
